File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long cc;

    do
    {
        cc = get_long("Number: ");
    }
    while (cc < 0);

    long sum = 0;

    // Iterate for up to 16 decimal places, going two at a time
    for (long i = 10; i < 100000000000000000; i *= 100)
    {
        // Get every other digit, starting with second to last
        long secondFromEnd = (cc / i) % 10;

        // Get every other digit, starting from last
        long firstFromEnd = (cc / (i / 10)) % 10;

        // Multiply secondFromEnd digits by 2
        long mult2 = secondFromEnd * 2;

        // Add mult2 product's digits to a sum, ie 18 = 1 + 8, 6 = 0 + 6
        // and then add firstFromEnd digits that are not multiplied
        sum += mult2 / 10 + mult2 % 10 + firstFromEnd;
    }

    // Check if checksum validates
    long isCardValid = sum % 10 == 0;

    if (isCardValid)
    {
        // AMEX 15 digits, starts with 34 or 37 - divide by 10000000000000
        if (cc / 10000000000000 == 34 || cc / 10000000000000 == 37)
        {
            printf("AMEX\n");
        }

        // MASTERCARD 16 digits, starts with 51, 52, 53, 54, 55 - divide by 100000000000000
        else if (cc / 100000000000000 == 51 ||
                 cc / 100000000000000 == 52 ||
                 cc / 100000000000000 == 53 ||
                 cc / 100000000000000 == 54 ||
                 cc / 100000000000000 == 55)
        {
            printf("MASTERCARD\n");
        }

        // VISA 13 or 16 digits, starts with 4 - divide by 1000000000000000 || 1000000000000
        else if (cc / 1000000000000000 == 4 || cc / 1000000000000 == 4)
        {
            printf("VISA\n");
        }

        // If none of the CC card starting patterns match, it's also invalid
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
# Refactor of pset1/credit
from cs50 import get_string


def main():
    card = get_string("Number: ")

    digit_sum = calculate_digit_sum(card)

    # If digit sum ends with a 0, it's potentially valid CC
    if digit_sum % 10 == 0:
        print(get_card_type(card))
    else:
        print("INVALID")


# Calculates the sum of the digits in the card
def calculate_digit_sum(card):
    card_len = len(card)
    is_even = False
    full_sum = 0

    # Iterate over each character in card number, going backwards
    for i in range(card_len - 1, -1,  -1):
        # cast character to an int
        digit = int(card[i])

        if is_even:
            full_sum += get_even_sum(digit)
        else:
            full_sum += digit

        is_even = not is_even

    return full_sum


# Calculates the sum of digits of even digits multiplied by 2
def get_even_sum(digit):
    return sum(int(mult_digit) for mult_digit in str(digit * 2))


# Returns a CC Card Type based on pattern
def get_card_type(card):
    # AMEX 15 digits, starts with 34 or 37
    if len(card) == 15 and int(card[0:2]) in [34, 37]:
        return "AMEX"

    # MASTERCARD 16 digits, starts with 51, 52, 53, 54, 55
    elif len(card) == 16 and int(card[0:2]) in [51, 52, 53, 54, 55]:
        return "MASTERCARD"

    # VISA 13 or 16 digits, starts with 4
    elif len(card) in [13, 16] and int(card[0]) == 4:
        return "VISA"

    # If none of the CC card starting patterns match, it's invalid
    else:
        return "INVALID"


if __name__ == "__main__":
    main()